src/lib/crypto/kimchi_backend/gadgets/keccak.ml

```diff
@@ -0,0 +1,720 @@
+    output
+
+  let _print (type f)
+      (module Circuit : Snarky_backendless.Snark_intf.Run with type field = f)
+      (state : Circuit.Field.t matrix) =
+    let open Circuit in
+    as_prover (fun () ->
+        for x = 0 to keccak_dim - 1 do
+          for y = 0 to keccak_dim - 1 do
+            let _elem =
+              Common.cvar_field_to_bignum_bigint_as_prover
+                (module Circuit)
+                state.(x).(y)
+            in
+            ()
+          done
+        done )
+end
+
+(* KECCAK HASH FUNCTION IMPLEMENTATION *)
+
+(* Performs the modulo operation, not remainder as in mod which instead computes integer remainder *)
+let modulo (x : int) (m : int) : int = ((x mod m) + m) mod m
+
+(* Pads a message M as:
+   * M || pad[x](|M|)
+   * Padding rule 0x06 ..0*..1.
+   * The padded message vector will start with the message vector
+   * followed by the 0*1 rule to fulfil a length that is a multiple of rate (in bytes)
+*)
+let pad_nist (type f)
+    (module Circuit : Snarky_backendless.Snark_intf.Run with type field = f)
+    (message : Circuit.Field.t list) (rate : int) : Circuit.Field.t list =
+  let open Circuit in
+  (* Find out desired length of the padding in bytes *)
+  (* If message is already rate bits, need to pad full rate again *)
+  let extra_bytes = (rate / 8) - (List.length message mod rate) in
+  (* 0x06 0x00 ... 0x00 0x80 or 0x86 *)
+  let last_field = Common.two_pow (module Circuit) 7 in
+  let last = Field.constant @@ Common.two_pow (module Circuit) 7 in
+  (* Create the padding vector *)
+  let pad = Array.create ~len:extra_bytes Field.zero in
+  pad.(0) <- Field.of_int 6 ;
+  pad.(extra_bytes - 1) <- Field.add pad.(extra_bytes - 1) last ;
+  Field.Assert.equal (Field.constant last_field) last ;
+  (* Cast the padding array to a list *)
+  let pad = Array.to_list pad in
+  (* Return the padded message *)
+  message @ pad
+
+(* Pads a message M as:
+   * M || pad[x](|M|)
+   * Padding rule 10*1.
+   * The padded message vector will start with the message vector
+   * followed by the 10*1 rule to fulfil a length that is a multiple of rate (in bytes)
+*)
+let pad_101 (type f)
+    (module Circuit : Snarky_backendless.Snark_intf.Run with type field = f)
+    (message : Circuit.Field.t list) (rate : int) : Circuit.Field.t list =
+  let open Circuit in
+  (* Find out desired length of the padding in bytes *)
+  (* If message is already rate bits, need to pad full rate again *)
+  let extra_bytes = (rate / 8) - (List.length message mod rate) in
+  (* 0x01 0x00 ... 0x00 0x80 or 0x81 *)
+  let last_field = Common.two_pow (module Circuit) 7 in
+  let last = Field.constant @@ last_field in
+  (* Create the padding vector *)
+  let pad = Array.create ~len:extra_bytes Field.zero in
+  pad.(0) <- Field.one ;
+  pad.(extra_bytes - 1) <- Field.add pad.(extra_bytes - 1) last ;
+  Field.Assert.equal (Field.constant last_field) last ;
+  (* Cast the padding array to a list *)
+  let pad = Array.to_list pad in
```

Comment:
Can we standardise on either `Array` or `List` in this module instead of mixing? Preferably `Array`